File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.py
# Last updated: 6/8/2025, 11:08:53 pm
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        st = []
        for i in range (len(prices)):
            while st and (prices[st[-1]]>=prices[i]):
                prices[st.pop()] -= prices[i]
            st.append(i)
        return prices
        

File: 1570-FinalPricesWithASpecialDiscountInAShop/1570-FinalPricesWithASpecialDiscountInAShop.java
// Last updated: 6/8/2025, 11:08:47 pm
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<prices.length;i++){
            while(!st.isEmpty() && prices[st.peek()]>=prices[i]){
                int idx = st.pop();
                prices[idx]-=prices[i];
            }
            st.push(i);
        }
        return prices;
    }
}